### ISSUE TITLE
Almost all fields marked as optional in the definitions

### ISSUE THREAD

**Issue by riccardo-forina** [syndesisio/syndesis#4085]
## This is a...

<pre><code>
[x] Feature request
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Documentation issue or request
</code></pre>

## Description

The definitions in the swagger.json that we currently generate have almost no field marked as required. E.g. the `Connector` definition defines only `actions` and `name` as required fields:

```json
"Connector" : {
    "type" : "object",
    "required" : [ "actions", "name" ],
    ...
}
```

In the UI we are trying to generate the Typescript typings from the swagger.json automatically. Fields that are not marked as required will be defined as optional by Typescript. For the above example, this is the interface we get:

```ts
export interface Connector {
    connectorGroup?: ConnectorGroup;
    description?: string;
    icon?: string;
    readonly uses?: OptionalInt;
    connectorFactory?: string;
    connectorCustomizers?: string[];
    connectorGroupId?: string;
    actionsSummary?: ActionsSummary;
    componentScheme?: string;
    id?: string;
    version?: number; // int32
    actions: ConnectorAction[];
    tags?: string[];
    name: string;
    properties?: {
        [name: string]: ConfigurationProperty;
    };
    configuredProperties?: {
        [name: string]: string;
    };
    dependencies?: Dependency[];
    metadata?: {
        [name: string]: string;
    };
}
```

Please note the `?` next to the property name, which is Typescript way to define an optional type.

The problem is that the fields aren't really optional (not all of them at least). This leads to very brittle code on the UI side, where we have to explicitly tell the compiler to ignore the potential `undefined` value. E.g.:

```jsx
<IntegrationsListItem
  integrationId={mi.integration.id!}
  integrationName={mi.integration.name}
  currentState={mi.integration!.currentState!}
  targetState={mi.integration!.targetState!}
  isConfigurationRequired={
    !!(
      mi.integration!.board!.warnings ||
      mi.integration!.board!.errors ||
      mi.integration!.board!.notices
    )
  }
  ...
>
```

Note all the `!` near a property name, which is Typescript way to say that the value is going to be valid.

Would it be possible to mark the fields as required - where it makes sense to do it of course - to better describe the shape of the data?

Input swagger.json: [swagger.json.zip](https://github.com/syndesisio/syndesis/files/2576837/swagger.json.zip)
Generated Typescript interfaces: [models.d.ts.zip](https://github.com/syndesisio/syndesis/files/2576836/models.d.ts.zip)

**Comment by KurtStam** [comment]
+1 

Note that in our API fields are required when annotated with @NotNull. I think in a lot of cases we can add @NotNull to variables that are not java.Optional. We've been lenient with this. This is likely a task that requires coordination between the front and backend teams.

**Comment by lburgazzoli** [comment]
I think this issue raises an additional - old but relevant - issue: we share the exact same model between frontend and backend and our API are almost a REST wrapper over the JsonDB.

If you look at the connector example Riccardo has posted, you'd notice that it contains information such as `connectorFactory`, `connectorCustomizers`, `connectorGroupId` and `dependencies` that are useless on the front end side but they go back and fort every time.

**Comment by gashcrumb** [comment]
Yeah, so I think we want to move towards having a shared contract between the backend and frontend, but not necessarily a shared model.  The contract should better represent the requirements of the interface designs that we get from UXD.

That being said it's a large undertaking :-)  I think the maybe the first pass could be to decide on what fields are absolutely required by both ends of the app.  @lburgazzoli can we avoid exposing these non-relevant fields for example?

**Comment by lburgazzoli** [comment]
Well we could theoretically do that but then we'll fall into another common problem: the integration we store into the db is what it is received from the UI which means that we have everything embedded (connections, connectors) for each step. 

At runtime we use such per step information to set-up the camel route and if we remove such fields we need to include an additional copy of the resource i.e. a connector and change the way we warm-up the runtime.

Another option is to remove such fields before sending a payload to the UI which is of course possible but as we use immutables for our model, we need to create an additional intermediate copy of an already huge object and then we need to re-add them once we receive the same payload from the UI which causes additional allocations.

So yeah, it is possible but not quick and easy :)

**Comment by gashcrumb** [comment]
> the integration we store into the db is what it is received from the UI

Here I'm reading "integration" as the JSON payload, please correct me if I'm wrong.  I think we probably want to avoid this, and I'm not sure I've proposed this.  I know in some cases in the UI we've requested having a combined object as a response so we could avoid making multiple requests to fetch all the data to satisfy a view especially where the required data for the view spans multiple objects.  

I would think this would be done by reference rather than storing direct copies in the DB, is this something that jsondb simply doesn't support at all?  If not then I guess any efforts to tidy up the interface need to wait until we decide what to do with the our current data store.

**Comment by lburgazzoli** [comment]
> > the integration we store into the db is what it is received from the UI
> 
> Here I'm reading "integration" as the JSON payload, please correct me if I'm wrong. I think we probably want to avoid this, and I'm not sure I've proposed this. I know in some cases in the UI we've requested having a combined object as a response so we could avoid making multiple requests to fetch all the data to satisfy a view especially where the required data for the view spans multiple objects.
>

Yes I meant the JSON payload and I think we were lazy or under pressure to release to work on a better and more efficient solution.
 
> I would think this would be done by reference rather than storing direct copies in the DB, is this something that jsondb simply doesn't support at all? If not then I guess any efforts to tidy up the interface need to wait until we decide what to do with the our current data store.

JsonDB do not support references or at least I'm not aware of that but we could simulate that. It is not so efficient because of all the immutables bits but it can be done. A possible issue is then doing a proper upgrade but that's another topic :)

**Comment by riccardo-forina** [comment]
I guess there is no out-of-the-box support for [json:api](https://jsonapi.org/), right?

**Comment by lburgazzoli** [comment]
don't think so
